File: 2015/13/main.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <algorithm>
#include <vector>

#include <chrono>

int get_index(const std::string& name) {
    if (name == "Alice") {
        return 0;
    } else if (name == "Bob") {
        return 1;
    } else if (name == "Carol") {
        return 2;
    } else if (name == "David") {
        return 3;
    } else if (name == "Eric") {
        return 4;
    } else if (name == "Frank") {
        return 5;
    } else if (name == "George") {
        return 6;
    } else if (name == "Mallory") {
        return 7;
    } else {
        return -1;
    }
}

int main() {
    auto start = std::chrono::high_resolution_clock::now();
    int happiness_matrix[9][9] = {0};
    int n = 9;

    std::fstream file("input.txt", std::ios::in);

    std::string line;
    while (std::getline(file, line)) {
        std::stringstream ss(line);
        std::string person, neighbour, loose_or_gain, tmp;
        int happiness;

        // Carol would lose 37 happiness units by sitting next to David.

        ss >> person >> tmp >> loose_or_gain >> happiness >> tmp >> tmp >> tmp >> tmp >> tmp >> tmp >> neighbour;

        if (loose_or_gain == "lose") {
            happiness = -happiness;
        }
        neighbour.pop_back();

        // std::cout << person << " " << neighbour << " " << happiness << std::endl;
        happiness_matrix[get_index(person)][get_index(neighbour)] = happiness;
    }

    int max_happiness = 0;
    std::vector<int> v;
    for (int i = 0; i < n; i++) {
        v.push_back(i);
    }

    do {
        int happiness = 0;
        for (int i = 0; i < n; i++) {
            happiness += happiness_matrix[v[i]][v[(i + 1) % n]];
            happiness += happiness_matrix[v[(i + 1) % n]][v[i]];
        }
        max_happiness = std::max(max_happiness, happiness);
    } while (std::next_permutation(v.begin(), v.end()));

    std::cout << max_happiness << std::endl;
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
    std::cout << "Execution time: " << duration << "ms" << std::endl;
    return 0;
}

File: 2015/13/main.py
import aiofiles
import asyncio

import numpy as np
import itertools

import time


async def main():
    t1 = time.time()
    
    person_index = {
        "Alice": 0,
        "Bob": 1,
        "Carol": 2,
        "David": 3,
        "Eric": 4,
        "Frank": 5,
        "George": 6,
        "Mallory": 7
    }
    happiness_matrix = np.zeros((9, 9))
    
    async with aiofiles.open("input.txt", mode="r") as f:
        async for line in f:
            data = line.split()
            
            person = data[0].strip()
            will_gain = True if data[2] == "gain" else False
            happiness = int(data[3]) if will_gain else -int(data[3])
            neighbour = data[10][:-1].strip()
            
            # happiness = -happiness
            
            # print(f"{person} {happiness} {neighbour}")
            happiness_matrix[person_index[person]][person_index[neighbour]] = happiness
    
    # normilze matrix to be positive
    print(happiness_matrix)
    
    total_change = brute_force(happiness_matrix)
    print(total_change)
    
    t2 = time.time()
    print(f"Execution time: {t2-t1}s")
    

def brute_force(matrix):
    max_happiness = 0
    n = len(matrix)
    for perm in itertools.permutations(range(n)):
        happiness = 0
        for i in range(n):
            happiness += matrix[perm[i]][perm[(i+1)%n]]
            happiness += matrix[perm[(i+1)%n]][perm[i]]
        
        max_happiness = max(max_happiness, happiness)
    return max_happiness


if __name__ == "__main__":
    asyncio.run(main())